### PR TITLE
Modularize stale endpoint gc in an independent cell

### PR DESCRIPTION
**Pull request by pippolo84**
Modularize the stale CiliumEndpoint GC that runs at startup in an independent cell. The GC is adapted to work with dependency injected resources and clientset, but it is functionally equivalent to the previous implementation (see the original PR for more details about the GC itself https://github.com/cilium/cilium/pull/20350).
The related test has also been adapted into a hive-based integration test.

This is a required step to migrate the k8s CEP watchers to Resource[T]. Refer to [this comment](https://github.com/cilium/cilium/issues/28159#issuecomment-1816498303) for more information.

Depends on ~https://github.com/cilium/cilium/pull/29243~ and ~https://github.com/cilium/cilium/pull/29244~

Related: https://github.com/cilium/cilium/issues/28159

### CODE REVIEW

**Comment by ti-mo** [comment]
@pippolo84 Please assign a CODEOWNER to `pkg/endpointgc`. :pray:

**Review by ti-mo — APPROVED** [review]

**Comment by pippolo84** [comment]
Kindly pinging @danehans @asauber @youngnick @dylandreimerink and @qmonnet for reviews, thanks in advance :pray:

**Review by dylandreimerink — APPROVED** [review]
The foundations bits looks good to me, not 100% familiar with the endpoints GC logic, will leave that to others.

**Review by qmonnet — APPROVED** [review]
Looks good, as far as I can tell. Apologies for the delay!

File: pkg/endpointgc/endpointgc.go
Comment (mhofstetter) [inline pkg/endpointgc/endpointgc.go]: nit: name alias `ciliumv2` or `cilium_v2`

File: pkg/endpointgc/endpointgc.go
Comment (mhofstetter) [inline pkg/endpointgc/endpointgc.go]: worth logging a message that endpoint GC is disabled?

File: pkg/endpointgc/cell.go
Comment (mhofstetter) [inline pkg/endpointgc/cell.go:32]: Field `EnableStaleCiliumEndpointCleanup` from the config is never written. IMO we need to point the second argument to the field. And set the default when value when initializing the `Config` 

```
cell.Config(Config{
  EnableStaleCiliumEndpointCleanup: true,
}),
```

```
flags.Bool("enable-stale-cilium-endpoint-cleanup", def.EnableStaleCiliumEndpointCleanup, "Enable running cleanup init procedure of local CiliumEndpoints which are not being managed.")
```

**Review by gandro — APPROVED** [review]
Looks good for my codeowners, thanks!

Unrelated to your refactor here, but looking at this code again, I think long-term we should reconsider the fact that we're using the Node IP as the node identifier. Using the node IP basically prohibits changing the node IP between agent restarts, since all CEPs with the old node IP will effectively be orphaned.

File: pkg/endpointgc/endpointgc.go
Comment (mhofstetter) [inline pkg/endpointgc/endpointgc.go]: is this an "actual" garbage collector? I wonder whether it would help to rename it (package, struct, filename, ...) to `(Stale)EndpointCleanup` - because it only runs once at the agent startup and we should avoid confusion with the actual EndpointGC from the operator.

**Review by mhofstetter — CHANGES_REQUESTED** [review]
nice, thanks @pippolo84! Some comments inline. ~~With one blocking issue that i assume that the config flag `enable-stale-cilium-endpoint-cleanup` is no longer working as expected.~~

**Review by mhofstetter — APPROVED** [review]
Nevermind, the flag seems to work as expected. only some non-blocking nits and one naming question left.

File: pkg/endpointgc/endpointgc.go
Comment (pippolo84) [reply pkg/endpointgc/endpointgc.go]: Changed the package name to `endpointcleanup` to avoid confusion (also changed the filenames, the struct and so on). :+1:

**Comment by pippolo84** [comment]
Removing some reviewers as their codeowner has been already covered.

**Review by christarazi — APPROVED** [review]

**Comment by julianwiedmann** [comment]
If this is a 1.15 release blocker, we now presumably also need a backport label. Adding it.

**Review by julianwiedmann — APPROVED** [review]

**Pull request merged**
